### PR TITLE
KV Watch Multiple Filters

### PR DESCRIPTION
**Pull request by scottf**
https://github.com/nats-io/nats-architecture-and-design/issues/277

### CODE REVIEW

File: src/NATS.Client/KeyValue/KeyValueWatchSubscription.cs [hunk 1]
```diff
@@ -52,7 +57,7 @@ public class KeyValueWatchSubscription : IDisposable
             else
             {
                 fromRevision = ConsumerConfiguration.UlongUnset; // easier on the builder since we aren't starting at a fromRevision
-                if (deliverPolicy == DeliverPolicy.New || kv._getLast(subscribeSubject) == null)
```
Comment (scottf) [inline src/NATS.Client/KeyValue/KeyValueWatchSubscription.cs:55]: This piece removed based on changes in java and discussions

File: src/NATS.Client/KeyValue/KeyValueWatchSubscription.cs [hunk 2]
```diff
@@ -91,7 +96,7 @@ void Handler(object sender, MsgHandlerEventArgs args)
                 }
             }
 
-            sub = kv.js.PushSubscribeAsync(subscribeSubject, Handler, false, pso);
+            sub = kv.js.PushSubscribeAsync(null, Handler, false, pso);
```
Comment (scottf) [inline src/NATS.Client/KeyValue/KeyValueWatchSubscription.cs:99]: filter subject is used when subscribe subject is null

File: src/Tests/IntegrationTests/TestKeyValue.cs [hunk 3]
```diff
@@ -915,6 +921,10 @@ public void TestWatch()
                 subs.Add(kv.Watch("key.*", starMetaWatcher, starMetaWatcher.WatchOptions));
                 subs.Add(kv.Watch("key.>", gtFullWatcher, gtFullWatcher.WatchOptions));
                 subs.Add(kv.Watch("key.>", gtMetaWatcher, gtMetaWatcher.WatchOptions));
+                subs.Add(kv.Watch(allKeys, multipleFullWatcher, multipleFullWatcher.WatchOptions));
+                subs.Add(kv.Watch(allKeys, multipleMetaWatcher, multipleMetaWatcher.WatchOptions));
```
Comment (scottf) [inline src/Tests/IntegrationTests/TestKeyValue.cs]: Tests the IList version

File: src/Tests/IntegrationTests/TestKeyValue.cs [hunk 4]
```diff
@@ -915,6 +921,10 @@ public void TestWatch()
...
+                subs.Add(kv.Watch(string.Join(",", allKeys), multipleFullWatcher2, multipleFullWatcher2.WatchOptions));
+                subs.Add(kv.Watch(string.Join(",", allKeys), multipleMetaWatcher2, multipleMetaWatcher2.WatchOptions));
```
Comment (scottf) [inline src/Tests/IntegrationTests/TestKeyValue.cs]: Test the comma delimited version

File: src/NATS.Client/KeyValue/KeyValue.cs [hunk 5]
```diff
@@ -175,28 +176,35 @@ public void Purge(string key, ulong expectedRevision)
         
         public KeyValueWatchSubscription Watch(string key, IKeyValueWatcher watcher, params KeyValueWatchOption[] watchOptions)
         {
-            Validator.ValidateKvKeyWildcardAllowedRequired(key);
-            Validator.ValidateNotNull(watcher, "Watcher is required");
-            return new KeyValueWatchSubscription(this, key, watcher, ConsumerConfiguration.UlongUnset, watchOptions);
+            return Watch(new List<string>(key.Split(',')), watcher, ConsumerConfiguration.UlongUnset, watchOptions);
```
Comment (mtmk) [inline src/NATS.Client/KeyValue/KeyValue.cs]: is comma delimited list safe? is comma not allowed as key?

File: src/NATS.Client/KeyValue/KeyValueWatchSubscription.cs (hunk 1 above)
Comment (mtmk) [reply src/NATS.Client/KeyValue/KeyValueWatchSubscription.cs:55]: is there a link?

**Review by mtmk — COMMENTED** [review]
only thing is I'm not sure about the comma delimited keys, otherwise looks good!

File: src/NATS.Client/KeyValue/KeyValue.cs (hunk 5 above)
Comment (scottf) [reply src/NATS.Client/KeyValue/KeyValue.cs]: Keys are limited and cannot have a comma
" must only contain A-Z, a-z, 0-9, '*', '-', '_', '/', '=', '>' or '.' and cannot start with '.' [" + s + "]"

File: src/NATS.Client/KeyValue/KeyValueWatchSubscription.cs (hunk 1 above)
Comment (scottf) [reply src/NATS.Client/KeyValue/KeyValueWatchSubscription.cs:55]: The same change was applied to java. The getLast code is actually brittle in the sense that a message could appear between the time the request is filled and the time the response reaches the client. It's also an edge case that was not implemented in go, so it's removed.

File: src/NATS.Client/KeyValue/KeyValue.cs (hunk 5 above)
Comment (scottf) [reply src/NATS.Client/KeyValue/KeyValue.cs]: See "Valid Keys" under https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-8.md#10

Comment (mtmk) [reply src/NATS.Client/KeyValue/KeyValue.cs]: is there a requirement to accept comma separated list of keys? i feel that could potentially be confusing

File: src/Tests/IntegrationTests/TestKeyValue.cs (hunk 4 above)
Comment (scottf) [reply src/Tests/IntegrationTests/TestKeyValue.cs]: removed

File: src/NATS.Client/KeyValue/KeyValue.cs (hunk 5 above)
Comment (scottf) [reply src/NATS.Client/KeyValue/KeyValue.cs]: removed this support

**Review by mtmk — APPROVED** [review]
LGTM

**Pull request merged**